source/compiler/qsc_eval/src/lib.rs

```diff
@@ -607,7 +607,10 @@ impl State {
     }
 
     fn push_scope(&mut self, env: &mut Env) {
-        env.push_scope(self.call_stack.len());
+        // `push_frame`, which increments the length of `self.call_stack` by 1,
+        // is called before `self.push_scope`.
+        // Since the first `frame_id` should be 0, we substract 1 here.
+        env.push_scope(self.call_stack.len() - 1);
```

Comment:
Ah, this is where the off-by-one error is coming from! Glad you found this, now the different env scopes will maintain a frame_id that matches the index of that frame in the call_stack.